File: apiendpoint/node/models/erpModels.js
// ====================================================================
// ERP SYSTEM - ADDITIONAL SEQUELIZE MODELS
// File: backend/node/models/erpModels.js
// Purpose: Additional models untuk melengkapi ERP system
// Date: 2025-08-27
// ====================================================================

const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

// ====================================================================
// INVENTORY & LOCATION MODELS
// ====================================================================

const InventoryLocation = sequelize.define('InventoryLocation', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    location_code: {
        type: DataTypes.STRING(50),
        unique: true,
        allowNull: false,
        comment: 'Location code (WH-A01, QC-01, etc)'
    },
    location_name: {
        type: DataTypes.STRING(100),
        allowNull: false,
        comment: 'Human readable location name'
    },
    location_type: {
        type: DataTypes.ENUM('raw_material', 'wip', 'finished_goods', 'quarantine', 'staging'),
        allowNull: false
    },
    warehouse_zone: {
        type: DataTypes.STRING(50),
        comment: 'Zone within warehouse'
    },
    capacity: {
        type: DataTypes.DECIMAL(12, 2),
        comment: 'Maximum capacity'
    },
    current_utilization: {
        type: DataTypes.DECIMAL(12, 2),
        defaultValue: 0.00,
        comment: 'Current stock level'
    },
    is_active: {
        type: DataTypes.BOOLEAN,
        defaultValue: true
    }
}, {
    tableName: 'inventory_locations',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    indexes: [
        { fields: ['location_type'] },
        { fields: ['is_active'] }
    ]
});

const InventoryMovement = sequelize.define('InventoryMovement', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    movement_number: {
        type: DataTypes.STRING(50),
        unique: true,
        allowNull: false,
        comment: 'Movement reference number'
    },
    part_number: {
        type: DataTypes.STRING(100),
        allowNull: false
    },
    movement_type: {
        type: DataTypes.ENUM('in', 'out', 'transfer', 'adjustment', 'scrap'),
        allowNull: false
    },
    from_location_id: {
        type: DataTypes.INTEGER,
        comment: 'Source location'
    },
    to_location_id: {
        type: DataTypes.INTEGER,
        comment: 'Destination location'
    },
    quantity: {
        type: DataTypes.DECIMAL(12, 3),
        allowNull: false
    },
    unit_cost: {
        type: DataTypes.DECIMAL(12, 2),
        comment: 'Unit cost at time of movement'
    },
    reference_type: {
        type: DataTypes.ENUM('production', 'delivery', 'receipt', 'adjustment', 'transfer', 'scrap'),
        allowNull: false
    },
    reference_id: {
        type: DataTypes.INTEGER,
        comment: 'Reference to source document'
    },
    reason_code: {
        type: DataTypes.STRING(50),
        comment: 'Reason for movement'
    },
    notes: {
        type: DataTypes.TEXT,
        comment: 'Additional notes'
    },
    user_id: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    movement_date: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW
    }
}, {
    tableName: 'inventory_movements',
    timestamps: false,
    indexes: [
        { fields: ['movement_date'] },
        { fields: ['movement_type'] },
        { fields: ['reference_type', 'reference_id'] }
    ]
});

const InventoryBalance = sequelize.define('InventoryBalance', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    part_number: {
        type: DataTypes.STRING(100),
        allowNull: false
    },
    location_id: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    available_quantity: {
        type: DataTypes.DECIMAL(12, 3),
        defaultValue: 0.00,
        comment: 'Available for use'
    },
    reserved_quantity: {
        type: DataTypes.DECIMAL(12, 3),
        defaultValue: 0.00,
        comment: 'Reserved for orders'
    },
    quarantine_quantity: {
        type: DataTypes.DECIMAL(12, 3),
        defaultValue: 0.00,
        comment: 'In quarantine'
    },
    average_cost: {
        type: DataTypes.DECIMAL(12, 2),
        defaultValue: 0.00,
        comment: 'Weighted average cost'
    },
    last_movement_date: {
        type: DataTypes.DATE
    },
    last_count_date: {
        type: DataTypes.DATE,
        comment: 'Last physical count'
    }
}, {
    tableName: 'inventory_balances',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    indexes: [
        { 
            fields: ['part_number', 'location_id'],
            unique: true,
            name: 'unique_part_location'
        },
        { fields: ['available_quantity'] }
    ]
});

// ====================================================================
// PRODUCTION PLANNING MODELS
// ====================================================================

const Machine = sequelize.define('Machine', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    machine_code: {
        type: DataTypes.STRING(50),
        unique: true,
        allowNull: false
    },
    machine_name: {
        type: DataTypes.STRING(100),
        allowNull: false
    },
    machine_type: {
        type: DataTypes.STRING(50),
        comment: 'CNC, Press, Assembly, etc'
    },
    location_id: {
        type: DataTypes.INTEGER,
        comment: 'Physical location'
    },
    capacity_per_hour: {
        type: DataTypes.DECIMAL(8, 2),
        comment: 'Standard capacity per hour'
    },
    status: {
        type: DataTypes.ENUM('active', 'maintenance', 'breakdown', 'idle'),
        defaultValue: 'active'
    },
    is_active: {
        type: DataTypes.BOOLEAN,
        defaultValue: true
    }
}, {
    tableName: 'machines',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    indexes: [
        { fields: ['status'] },
        { fields: ['machine_type'] }
    ]
});

const ProductionSchedule = sequelize.define('ProductionSchedule', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    schedule_number: {
        type: DataTypes.STRING(50),
        unique: true,
        allowNull: false
    },
    production_order_id: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    machine_id: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    operator_id: {
        type: DataTypes.INTEGER,
        comment: 'Assigned operator'
    },
    scheduled_start: {
        type: DataTypes.DATE,
        allowNull: false
    },
    scheduled_end: {
        type: DataTypes.DATE,
        allowNull: false
    },
    actual_start: {
        type: DataTypes.DATE
    },
    actual_end: {
        type: DataTypes.DATE
    },
    status: {
        type: DataTypes.ENUM('scheduled', 'in_progress', 'completed', 'cancelled', 'on_hold'),
        defaultValue: 'scheduled'
    },
    priority: {
        type: DataTypes.INTEGER,
        defaultValue: 5,
        comment: '1=Highest, 10=Lowest'
    },
    setup_time_minutes: {
        type: DataTypes.INTEGER,
        defaultValue: 0
    },
    estimated_runtime_minutes: {
        type: DataTypes.INTEGER
    },
    actual_runtime_minutes: {
        type: DataTypes.INTEGER
    },
    notes: {
        type: DataTypes.TEXT
    },
    created_by: {
        type: DataTypes.INTEGER,
        allowNull: false
    }
}, {
    tableName: 'production_schedules',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    indexes: [
        { fields: ['scheduled_start', 'scheduled_end'] },
        { fields: ['status'] },
        { fields: ['machine_id', 'scheduled_start'] }
    ]
});

const BillOfMaterials = sequelize.define('BillOfMaterials', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    parent_part_number: {
        type: DataTypes.STRING(100),
        allowNull: false,
        comment: 'Finished product'
    },
    child_part_number: {
        type: DataTypes.STRING(100),
        allowNull: false,
        comment: 'Component/material'
    },
    quantity_required: {
        type: DataTypes.DECIMAL(12, 4),
        allowNull: false,
        comment: 'Quantity needed per parent unit'
    },
    unit_of_measure: {
        type: DataTypes.STRING(20),
        defaultValue: 'PCS'
    },
    scrap_factor: {
        type: DataTypes.DECIMAL(5, 4),
        defaultValue: 0.0000,
        comment: 'Expected scrap percentage'
    },
    operation_sequence: {
        type: DataTypes.INTEGER,
        defaultValue: 1,
        comment: 'When this component is used'
    },
    is_active: {
        type: DataTypes.BOOLEAN,
        defaultValue: true
    },
    effective_date: {
        type: DataTypes.DATEONLY,
        allowNull: false
    },
    expiry_date: {
        type: DataTypes.DATEONLY
    }
}, {
    tableName: 'bill_of_materials',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    indexes: [
        { 
            fields: ['parent_part_number', 'child_part_number', 'effective_date'],
            unique: true,
            name: 'unique_bom_component'
        },
        { fields: ['parent_part_number'] },
        { fields: ['child_part_number'] },
        { fields: ['is_active', 'effective_date', 'expiry_date'] }
    ]
});

// ====================================================================
// PROCUREMENT MODELS
// ====================================================================

const Supplier = sequelize.define('Supplier', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    supplier_code: {
        type: DataTypes.STRING(50),
        unique: true,
        allowNull: false
    },
    supplier_name: {
        type: DataTypes.STRING(200),
        allowNull: false
    },
    contact_person: {
        type: DataTypes.STRING(100)
    },
    email: {
        type: DataTypes.STRING(100)
    },
    phone: {
        type: DataTypes.STRING(50)
    },
    address: {
        type: DataTypes.TEXT
    },
    city: {
        type: DataTypes.STRING(100)
    },
    country: {
        type: DataTypes.STRING(100),
        defaultValue: 'Indonesia'
    },
    tax_number: {
        type: DataTypes.STRING(50)
    },
    payment_terms: {
        type: DataTypes.STRING(100),
        comment: 'NET 30, COD, etc'
    },
    currency: {
        type: DataTypes.STRING(10),
        defaultValue: 'IDR'
    },
    is_active: {
        type: DataTypes.BOOLEAN,
        defaultValue: true
    },
    rating: {
        type: DataTypes.DECIMAL(3, 2),
        comment: 'Supplier rating 1-5'
    }
}, {
    tableName: 'suppliers',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    indexes: [
        { fields: ['supplier_name'] },
        { fields: ['is_active'] }
    ]
});

const PurchaseOrder = sequelize.define('PurchaseOrder', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    po_number: {
        type: DataTypes.STRING(50),
        unique: true,
        allowNull: false
    },
    supplier_id: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    po_date: {
        type: DataTypes.DATEONLY,
        allowNull: false
    },
    expected_delivery_date: {
        type: DataTypes.DATEONLY
    },
    status: {
        type: DataTypes.ENUM('draft', 'sent', 'acknowledged', 'partial_received', 'completed', 'cancelled'),
        defaultValue: 'draft'
    },
    total_amount: {
        type: DataTypes.DECIMAL(15, 2)
    },
    currency: {
        type: DataTypes.STRING(10),
        defaultValue: 'IDR'
    },
    payment_terms: {
        type: DataTypes.STRING(100)
    },
    delivery_address: {
        type: DataTypes.TEXT
    },
    notes: {
        type: DataTypes.TEXT
    },
    created_by: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    approved_by: {
        type: DataTypes.INTEGER
    },
    approved_at: {
        type: DataTypes.DATE
    }
}, {
    tableName: 'purchase_orders',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    indexes: [
        { fields: ['po_date'] },
        { fields: ['status'] },
        { fields: ['supplier_id', 'po_date'] }
    ]
});

// ====================================================================
// SYSTEM MANAGEMENT MODELS
// ====================================================================

const IntegrationQueue = sequelize.define('IntegrationQueue', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    queue_number: {
        type: DataTypes.STRING(50),
        unique: true,
        allowNull: false
    },
    system_name: {
        type: DataTypes.STRING(50),
        allowNull: false,
        comment: 'Target system: WMS, ERP, QC, etc'
    },
    action_type: {
        type: DataTypes.STRING(50),
        allowNull: false,
        comment: 'create, update, delete, sync'
    },
    entity_type: {
        type: DataTypes.STRING(50),
        allowNull: false,
        comment: 'production_order, delivery, etc'
    },
    entity_id: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    payload: {
        type: DataTypes.JSON,
        comment: 'Data to be sent'
    },
    status: {
        type: DataTypes.ENUM('pending', 'processing', 'completed', 'failed', 'retry'),
        defaultValue: 'pending'
    },
    priority: {
        type: DataTypes.INTEGER,
        defaultValue: 5
    },
    retry_count: {
        type: DataTypes.INTEGER,
        defaultValue: 0
    },
    max_retries: {
        type: DataTypes.INTEGER,
        defaultValue: 3
    },
    error_message: {
        type: DataTypes.TEXT
    },
    response_data: {
        type: DataTypes.JSON
    },
    processed_at: {
        type: DataTypes.DATE
    },
    next_retry_at: {
        type: DataTypes.DATE
    }
}, {
    tableName: 'integration_queue',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: false,
    indexes: [
        { fields: ['status'] },
        { fields: ['system_name', 'status'] },
        { fields: ['next_retry_at'] }
    ]
});

const WorkflowState = sequelize.define('WorkflowState', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    entity_type: {
        type: DataTypes.STRING(50),
        allowNull: false,
        comment: 'production_order, delivery, etc'
    },
    entity_id: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    workflow_name: {
        type: DataTypes.STRING(100),
        allowNull: false,
        comment: 'production_workflow, delivery_workflow'
    },
    current_state: {
        type: DataTypes.STRING(50),
        allowNull: false
    },
    previous_state: {
        type: DataTypes.STRING(50)
    },
    next_possible_states: {
        type: DataTypes.JSON,
        comment: 'Array of possible next states'
    },
    state_data: {
        type: DataTypes.JSON,
        comment: 'Additional state-specific data'
    },
    changed_by: {
        type: DataTypes.INTEGER,
        allowNull: false
    },
    changed_at: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW
    },
    notes: {
        type: DataTypes.TEXT
    },
    is_active: {
        type: DataTypes.BOOLEAN,
        defaultValue: true
    }
}, {
    tableName: 'workflow_states',
    timestamps: false,
    indexes: [
        { 
            fields: ['entity_type', 'entity_id', 'workflow_name', 'is_active'],
            unique: true,
            name: 'unique_entity_workflow'
        },
        { fields: ['workflow_name', 'current_state'] },
        { fields: ['entity_type', 'entity_id'] }
    ]
});

const SystemConfiguration = sequelize.define('SystemConfiguration', {
    id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true
    },
    config_key: {
        type: DataTypes.STRING(100),
        unique: true,
        allowNull: false
    },
    config_value: {
        type: DataTypes.TEXT,
        allowNull: false
    },
    config_type: {
        type: DataTypes.ENUM('string', 'number', 'boolean', 'json'),
        defaultValue: 'string'
    },
    description: {
        type: DataTypes.TEXT
    },
    is_editable: {
        type: DataTypes.BOOLEAN,
        defaultValue: true
    },
    category: {
        type: DataTypes.STRING(50),
        defaultValue: 'general'
    }
}, {
    tableName: 'system_configuration',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    indexes: [
        { fields: ['category'] }
    ]
});

// ====================================================================
// MODEL ASSOCIATIONS
// ====================================================================

// Note: Associations will be setup in index.js after all models are loaded
// to avoid circular dependency issues

module.exports = {
    // Inventory & Location
    InventoryLocation,
    InventoryMovement,
    InventoryBalance,
    
    // Production Planning
    Machine,
    ProductionSchedule,
    BillOfMaterials,
    
    // Procurement
    Supplier,
    PurchaseOrder,
    
    // System Management
    IntegrationQueue,
    WorkflowState,
    SystemConfiguration
};

File: apiendpoint/app/models/erp_models.py
"""
ERP Database Models for Query Operations
Mapped to the same database tables used by Node.js Command Service
Read-only access for reporting and analytics
"""
from sqlalchemy import Column, Integer, String, DateTime, Text, Boolean, ForeignKey, Date, Time, Numeric
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship
from datetime import datetime, date

Base = declarative_base()

# Core ERP Tables
class InventoryLocation(Base):
    __tablename__ = "inventory_locations"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    location_code = Column(String(50), unique=True, nullable=False, index=True)  # Database has varchar(50)
    location_name = Column(String(100), nullable=False)
    location_type = Column(String(20), nullable=False)  # enum in database
    warehouse_zone = Column(String(50))  # Database has 'warehouse_zone', not 'warehouse_section'
    capacity = Column(Numeric(12, 2))  # New field in database
    current_utilization = Column(Numeric(12, 2), default=0.00)  # New field in database
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class InventoryBalance(Base):
    __tablename__ = "inventory_balances"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    part_number = Column(String(100), nullable=False, index=True)  # Database has varchar(100)
    location_id = Column(Integer, ForeignKey("inventory_locations.id"), nullable=False)
    available_quantity = Column(Numeric(12, 3), default=0.000)  # Match database precision
    reserved_quantity = Column(Numeric(12, 3), default=0.000)
    quarantine_quantity = Column(Numeric(12, 3), default=0.000)  # New field in database
    average_cost = Column(Numeric(12, 2), default=0.00)  # Match database precision
    last_movement_date = Column(DateTime)
    last_count_date = Column(DateTime)  # New field in database
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    location = relationship("InventoryLocation", backref="balances")

class InventoryMovement(Base):
    __tablename__ = "inventory_movements"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    movement_number = Column(String(50), unique=True, nullable=False, index=True)  # Database has varchar(50)
    part_number = Column(String(100), nullable=False, index=True)  # Database has varchar(100)
    movement_type = Column(String(20), nullable=False)  # ENUM: in, out, transfer, adjustment, scrap
    from_location_id = Column(Integer, ForeignKey("inventory_locations.id"))
    to_location_id = Column(Integer, ForeignKey("inventory_locations.id"))
    quantity = Column(Numeric(12, 3), nullable=False)  # Database has decimal(12,3)
    unit_cost = Column(Numeric(12, 2))  # Database has decimal(12,2)
    reference_type = Column(String(20))  # ENUM: production, delivery, receipt, adjustment, transfer, scrap
    reference_id = Column(Integer)
    reason_code = Column(String(50))  # Database has varchar(50)
    notes = Column(Text)
    user_id = Column(Integer)
    movement_date = Column(DateTime)  # Database has timestamp, no default needed
    # created_at = Column(DateTime, default=datetime.utcnow)  # Column doesn't exist in database
    
    # Relationships
    from_location = relationship("InventoryLocation", foreign_keys=[from_location_id])
    to_location = relationship("InventoryLocation", foreign_keys=[to_location_id])

class Machine(Base):
    __tablename__ = "machines"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    machine_code = Column(String(50), nullable=False, index=True)
    machine_name = Column(String(100), nullable=False)
    machine_type = Column(String(50))
    location_id = Column(Integer)  # Database has 'location_id', not 'location'
    capacity_per_hour = Column(Numeric(8, 2))
    status = Column(String(20), default='active')  # Database has ENUM
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class Supplier(Base):
    __tablename__ = "suppliers"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    supplier_code = Column(String(20), unique=True, nullable=False, index=True)
    supplier_name = Column(String(100), nullable=False)
    contact_person = Column(String(100))
    phone = Column(String(20))
    email = Column(String(100))
    address = Column(Text)
    payment_terms = Column(String(50))
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class PurchaseOrder(Base):
    __tablename__ = "purchase_orders"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    po_number = Column(String(30), unique=True, nullable=False, index=True)
    supplier_id = Column(Integer, ForeignKey("suppliers.id"), nullable=False)
    order_date = Column(Date, nullable=False)
    expected_delivery = Column(Date)
    status = Column(String(20), default='draft')  # draft, sent, confirmed, received, closed
    total_amount = Column(Numeric(15, 2))
    notes = Column(Text)
    created_by = Column(Integer)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    supplier = relationship("Supplier", backref="purchase_orders")

class PurchaseOrderItem(Base):
    __tablename__ = "purchase_order_items"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    po_id = Column(Integer, ForeignKey("purchase_orders.id"), nullable=False)
    part_number = Column(String(50), nullable=False)
    description = Column(String(200))
    quantity_ordered = Column(Numeric(15, 3), nullable=False)
    quantity_received = Column(Numeric(15, 3), default=0)
    unit_price = Column(Numeric(15, 4), nullable=False)
    total_price = Column(Numeric(15, 2))
    expected_date = Column(Date)
    created_at = Column(DateTime, default=datetime.utcnow)
    
    # Relationships
    purchase_order = relationship("PurchaseOrder", backref="items")

class BillOfMaterials(Base):
    __tablename__ = "bill_of_materials"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    parent_part_number = Column(String(50), nullable=False, index=True)
    child_part_number = Column(String(50), nullable=False, index=True)
    quantity_required = Column(Numeric(15, 6), nullable=False)
    unit_of_measure = Column(String(10), nullable=False)
    scrap_factor = Column(Numeric(5, 4), default=0)
    operation_sequence = Column(Integer, default=1)
    effective_date = Column(Date, nullable=False)
    expiry_date = Column(Date)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class ProductionSchedule(Base):
    __tablename__ = "production_schedules"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    schedule_number = Column(String(30), unique=True, nullable=False, index=True)
    production_order_id = Column(Integer, nullable=False, index=True)
    machine_id = Column(Integer, ForeignKey("machines.id"), nullable=False)
    operator_id = Column(Integer)
    scheduled_start = Column(DateTime, nullable=False)
    scheduled_end = Column(DateTime, nullable=False)
    actual_start = Column(DateTime)
    actual_end = Column(DateTime)
    status = Column(String(20), default='scheduled')  # scheduled, started, completed, cancelled
    priority = Column(Integer, default=5)
    estimated_runtime_minutes = Column(Integer)
    actual_runtime_minutes = Column(Integer)
    # efficiency_percentage = Column(Numeric(5, 2))  # Column doesn't exist in database
    notes = Column(Text)
    created_by = Column(Integer)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    machine = relationship("Machine", backref="schedules")

class WorkflowState(Base):
    __tablename__ = "workflow_states"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    entity_type = Column(String(50), nullable=False, index=True)  # production_order, quality_check, etc
    entity_id = Column(Integer, nullable=False, index=True)
    current_state = Column(String(50), nullable=False)
    previous_state = Column(String(50))
    state_data = Column(Text)  # JSON data for additional state information
    changed_by = Column(Integer)
    changed_at = Column(DateTime, default=datetime.utcnow)
    notes = Column(Text)

class User(Base):
    __tablename__ = "users"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    email = Column(String(100), unique=True, nullable=False, index=True)
    password = Column(String(255), nullable=False)
    name = Column(String(100), nullable=False)
    role = Column(String(20), nullable=False)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class ProductionOrder(Base):
    __tablename__ = "production_orders"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    job_order = Column(String(50), nullable=False, index=True)  # Database has varchar(50)
    part_number = Column(String(100), nullable=False, index=True)  # Database has varchar(100)
    plan_quantity = Column(Numeric(12, 2), nullable=False)  # Database has decimal(12,2)
    machine_name = Column(String(100))  # This field exists in database
    start_date = Column(Date, nullable=False)
    status = Column(String(20), default='running')  # Database has ENUM with 'running' as default
    workflow_status = Column(String(20), default='planning')  # Database has this field
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class OutputMc(Base):
    __tablename__ = "output_mc"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    job_order = Column(String(30), nullable=False, index=True)
    operation_date = Column(Date, nullable=False)
    shift = Column(String(10), nullable=False)
    part_number = Column(String(50), nullable=False)
    part_name = Column(String(200))
    machine_name = Column(String(100))
    actual_quantity = Column(Numeric(10, 2), nullable=False)
    ng_quantity = Column(Numeric(10, 2), default=0)
    operator_name = Column(String(100))
    notes = Column(Text)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

class MasterProd(Base):
    __tablename__ = "master_prod"
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    part_number = Column(String(100), unique=True, nullable=False, index=True)  # Database has varchar(100)
    description = Column(Text)  # Database has 'description', not 'part_name'
    unit_of_measure = Column(String(20), default='PCS')  # Database has varchar(20)
    standard_cost = Column(Numeric(12, 2), default=0.00)  # Database has decimal(12,2)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
